libraries/AP_OpenDroneID/AP_OpenDroneID.cpp

```diff
@@ -0,0 +1,546 @@
+#include <AP_AHRS/AP_AHRS.h>
+
+const AP_Param::GroupInfo AP_OpenDroneID::var_info[] = {
+    // @Param: SER_NUM
+    // @DisplayName: Serial Number
+    // @Description: Serial Number
+    // @User: Advanced
+    AP_GROUPINFO("SER_NUM", 1, AP_OpenDroneID, _serial_number, 0),
+
+    AP_GROUPEND
+};
+
+// constructor
+AP_OpenDroneID::AP_OpenDroneID()
+{
+#if CONFIG_HAL_BOARD == HAL_BOARD_SITL
+    if (_singleton != nullptr) {
+        AP_HAL::panic("OpenDroneID must be singleton");
+    }
+#endif
+    _singleton = this;
+    AP_Param::setup_object_defaults(this, var_info);
+}
+
+void AP_OpenDroneID::init()
+{
+    _last_send_dynamic_messages_ms = 0;
+    _last_send_static_messages_ms = 0;
+    _mavlink_dynamic_period_ms = 1000;
+    _mavlink_static_period_ms = 10000;
+
+     _chan = MAVLINK_COMM_0; //set default to MAVLink channel 0
+     _id_type = MAV_ODID_ID_TYPE_NONE;
+     _ua_type = MAV_ODID_UA_TYPE_NONE;
+     _uav_status = MAV_ODID_STATUS_UNDECLARED;
+    _height_reference = MAV_ODID_HEIGHT_REF_OVER_TAKEOFF;
+
+    _operator_location_type = MAV_ODID_OPERATOR_LOCATION_TYPE_TAKEOFF;
+    _classification_type = MAV_ODID_CLASSIFICATION_TYPE_UNDECLARED;
+    _category_eu = MAV_ODID_CATEGORY_EU_UNDECLARED;
+    _class_eu = MAV_ODID_CLASS_EU_UNDECLARED;
+    _operator_id_type = MAV_ODID_OPERATOR_ID_TYPE_CAA;
+    _description_type = MAV_ODID_DESC_TYPE_TEXT;
+    _operator_timestamp = 0;
+    _operator_position.lat = 0;
+    _operator_position.lng = 0;
+    _operator_position.alt = (int32_t) -1000*100; //default -1000m
+
+    _area_count = 1; //default 1
+    _area_radius = 0; //default 0
+    _area_ceiling = -1000; //default -1000
+    _area_floor = -1000; //default -1000
+
+    memset(_id_or_mac,0,ODID_ID_SIZE);
+    memset(_uas_id,0,ODID_ID_SIZE + 1);
+    memset(_operator_id,0,ODID_ID_SIZE + 1);
+    memset(_description,0,ODID_STR_SIZE + 1);
+
+    transmit_basic_ID_message = false; //default we don't send any packets
+    transmit_location_message = false;
+    transmit_self_ID_message = false;
+    transmit_system_message = false;
+    transmit_operator_id_message = false;
+
+
+//code to be replaced: set dummy values in open drone ID messages
+    set_mavlink_channel(MAVLINK_COMM_1); //set to MAVLink channel 1
+    set_mavlink_dynamic_messages_period_ms(1000); //send dynamic messages at 1 Hz frequency i.e. every 1 second
+    set_mavlink_static_messages_period_ms(10000); //send dynamic messages at 0.1 Hz frequency i.e. every 10 seconds
+    set_id_type(MAV_ODID_ID_TYPE_SERIAL_NUMBER);
+    set_ua_type(MAV_ODID_UA_TYPE_AEROPLANE);
+
+    set_operator_location_type(MAV_ODID_OPERATOR_LOCATION_TYPE_LIVE_GNSS);
+    set_classification_type(MAV_ODID_CLASSIFICATION_TYPE_UNDECLARED);
+    set_category_eu(MAV_ODID_CATEGORY_EU_UNDECLARED);
+    set_class_eu(MAV_ODID_CLASS_EU_UNDECLARED);
+    set_operator_id_type(MAV_ODID_OPERATOR_ID_TYPE_CAA);
+    set_description_type(MAV_ODID_DESC_TYPE_TEXT);
+
+    set_operator_latitude(10);
```

Comment:
Is is possible for Ardupilot to provide this information?

If not, this should be set to the take-off location.

Same for set_operator_altitude_geo below.

Reply:
fixed to come from GCS